### PR TITLE
Better nano-zlib implementation.

### PR DESCRIPTION
**Pull request by erikcorry**
It stil only does run-length encoding
for speed and memory savings, but now
it is capable of recognizing repeating
patterns of 2, 3, or 4 bytes as seen in
color image files.  Big improvement in
very simple color PNG files.

### CODE REVIEW

File: src/nano_zlib.cc [hunk 1]
```diff
@@ -45,18 +45,128 @@ word ZlibRle::add(const uint8* contents, intptr_t extra) {
...
+        last_bytes_valid_ = 4;
+        unemitted_bytes_ = 0;
+        unemitted_bytes_valid_ = 0;
+      } else if (last_bytes_valid_ >= 2 &&
+          (last_bytes_ & 0xffff) == (unemitted_bytes_ & 0xffff) &&
+          (unemitted_bytes_ >> 16) == (unemitted_bytes_ & 0xffff)) {
+        // We recognized that the four unemitted bytes are two repetitions of
+        // the last two emitted ones.
+        mode_ = REP2;
+        repetitions_ = 4;
```
Comment (floitsch) [inline src/nano_zlib.cc:74]: Feels a bit wrong to call it "repetitions = 4" if it's only two repetitions.

File: src/nano_zlib.cc [hunk 2]
```diff
@@ -45,18 +45,128 @@ word ZlibRle::add(const uint8* contents, intptr_t extra) {
...
+        last_bytes_valid_ = 4;
+        unemitted_bytes_ &= 0xff;
+        unemitted_bytes_valid_ = 1;
+        output_repetitions();  // Will flush completely since repetitions_ == 3.
+      } else if (last_bytes_valid_ >= 4 &&
+          last_bytes_ == unemitted_bytes_) {
+        // We recognized that the four unemitted bytes are the same as
+        // the last four emitted ones.  Start a run with offset 4.
+        mode_ = REP4;
+        repetitions_ = 4;
```
Comment (floitsch) [inline src/nano_zlib.cc:99]: Same: it's just one repetition of 4 bytes. Not 4 of them.

File: src/nano_zlib.cc [hunk 3]
```diff
@@ -45,18 +45,128 @@ word ZlibRle::add(const uint8* contents, intptr_t extra) {
...
+        unemitted_bytes_ = 0;
+        unemitted_bytes_valid_ = 0;
+      } else if (last_bytes_valid_ >= 3 &&
+          (last_bytes_ & 0xffffff) == (unemitted_bytes_ >> 8)) {
+        // We recognized that the first four of the unemitted bytes are
```
Comment (floitsch) [inline src/nano_zlib.cc:104]:
```suggestion
        // We recognized that the first three of the unemitted bytes are
```

File: src/nano_zlib.cc [hunk 4]
```diff
@@ -45,18 +45,128 @@ word ZlibRle::add(const uint8* contents, intptr_t extra) {
...
+        // the same as the last three emitted ones.
+        mode_ = REP3;
+        if ((unemitted_bytes_ & 0xff) == ((last_bytes_ >> 16) & 0xff)) {
```
Comment (floitsch) [inline src/nano_zlib.cc:107]: Why the shift of 16?

File: src/nano_zlib.cc [hunk 5]
```diff
@@ -45,18 +45,128 @@ word ZlibRle::add(const uint8* contents, intptr_t extra) {
...
+          // The last unemitted byte also matches the pattern of three
```
Comment (floitsch) [inline src/nano_zlib.cc:108]: Not clear which pattern we are talking about.

File: src/nano_zlib.cc [hunk 6]
```diff
@@ -45,18 +45,128 @@ word ZlibRle::add(const uint8* contents, intptr_t extra) {
...
+        unemitted_bytes_ |= b;
+      }
     } else {
-      while (repetitions_ != 0) output_repetitions();
-      last_byte_ = b;
-      literal(b);
+      // We have some sort of run going.  As long as it is running we don't
+      // have unemitted bytes.
+      ASSERT(unemitted_bytes_valid_ == 0);
+      int shift = static_cast<int>(mode_ - 1) << 3;
```
Comment (floitsch) [inline src/nano_zlib.cc:149]: Add explicit values to the Mode constants and add a comment explaining that their value is crucial.
Alternatively make a map from mode to shift. (Almost prefer that).

File: src/nano_zlib.cc [hunk 7]
```diff
@@ -45,18 +45,128 @@ word ZlibRle::add(const uint8* contents, intptr_t extra) {
...
+        unemitted_bytes_ <<= 8;
+        unemitted_bytes_valid_++;
+        unemitted_bytes_ |= b;
+      }
     } else {
-      while (repetitions_ != 0) output_repetitions();
-      last_byte_ = b;
-      literal(b);
+      // We have some sort of run going.  As long as it is running we don't
+      // have unemitted bytes.
```
Comment (floitsch) [inline src/nano_zlib.cc:147]: Do I understand this correctly, that the "mode" defines the size of the repeated bytes. Or (in other words) how far back each byte needs to look to find itself again?
If yes, mention that somewhere.

**Review by floitsch — APPROVED** [review]
LGTM.

File: src/nano_zlib.cc (hunk 2 above)
Comment (erikcorry) [reply src/nano_zlib.cc]: Renamed to bytes_repeated_

File: src/nano_zlib.cc (hunk 4 above)
Comment (erikcorry) [reply src/nano_zlib.cc:107]: This should be clearer now with the comment.  We have recently emitted the pattern abc and we are checking for the unemitted bytes abca.  The shift of 16 gets the a down to the low bits.

**Pull request merged**